python/instrumentation/openinference-instrumentation-dspy/src/openinference/instrumentation/dspy/__init__.py

```diff
@@ -0,0 +1,169 @@
+
+        # Instrument LM (language model) calls
+        from dsp.modules.lm import LM
+
+        language_model_classes = LM.__subclasses__()
+        for lm in language_model_classes:
+            wrap_function_wrapper(
+                module=_DSP_MODULE,
+                name=lm.__name__ + ".basic_request",
+                wrapper=_LMBasicRequestWrapper(tracer),
+            )
+
+        # Instrument DSPy constructs
+        wrap_function_wrapper(
+            module=_DSPY_MODULE,
+            name="Predict.forward",
+            wrapper=_PredictForwardWrapper(tracer),
+        )
+
+    def _uninstrument(self, **kwargs: Any) -> None:
+        from dsp.modules.lm import LM
+
+        language_model_classes = LM.__subclasses__()
+        for lm in language_model_classes:
+            if hasattr(lm.request, "__wrapped__"):
+                lm.request = lm.request.__wrapped__
+
+        # Restore DSPy constructs
+        from dspy import Predict
+
+        if hasattr(Predict.forward, "__wrapped__"):
+            Predict.forward = Predict.forward.__wrapped__
+
+
+class _WithTracer(ABC):
+    """
+    Base class for wrappers that need a tracer. Acts as a trait for the wrappers
+    """
+
+    def __init__(self, tracer: trace_api.Tracer, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self._tracer = tracer
+
+
+class _LMBasicRequestWrapper(_WithTracer):
+    """
+    Wrapper for DSP LM.basic_request
+    Captures all calls to language models (lm)
+    """
+
+    def __call__(
+        self,
+        wrapped: Callable[..., Any],
+        instance: Any,
+        args: Tuple[type, Any],
+        kwargs: Mapping[str, Any],
+    ) -> Any:
+        prompt = args[0]
+        invocation_parameters = {**instance.kwargs, **kwargs}
+        span_name = instance.__class__.__name__ + ".request"
+        with self._tracer.start_as_current_span(
+            span_name,
+            attributes={
+                SpanAttributes.OPENINFERENCE_SPAN_KIND: OpenInferenceSpanKindValues.LLM.value,
+                SpanAttributes.LLM_MODEL_NAME: instance.kwargs.get("model"),
+                SpanAttributes.LLM_INVOCATION_PARAMETERS: json.dumps(invocation_parameters),
+                SpanAttributes.INPUT_VALUE: str(prompt),
+                SpanAttributes.INPUT_MIME_TYPE: OpenInferenceMimeTypeValues.TEXT.value,
+            },
+        ) as span:
+            try:
+                response = wrapped(*args, **kwargs)
+            except Exception as exception:
+                span.set_status(trace_api.Status(trace_api.StatusCode.ERROR, str(exception)))
+                span.record_exception(exception)
+                raise
+            # TODO: parse usage. Need to decide if this
+            # instrumentation should be used in conjunction model instrumentation
+            span.set_attributes(
+                {
+                    SpanAttributes.OUTPUT_VALUE: json.dumps(response),
```

Comment:
Did we want to switch to having the output field be the string of the output message for readability inside Phoenix?

Reply:
Yeah this is def. an area of improvement. I'm struggling with whether or not this tracing will be mounted with an openAI instrumentation at the same time. Let's leave it for now but agree it's not ideal.

What we also could do is parse the output messages but I punted on that too.